### PR TITLE
Split `omitUrlParams` in `lib/url`

### PR DESCRIPTION
**Pull request by sgomes**
This is one of several PRs aiming at splitting up `lib/url` into several files for tree-shaking, replacing usage of the browserified node `url` module with native functionality, and generally improving it.

This PR focuses on `omitUrlParams`, which I'm splitting into its own module and reimplementing using native functionality, dropping both `url` and `lodash`.

This PR was split out from #36531, which got too big.

#### Changes proposed in this Pull Request

* Split out `omitUrlParams` in `lib/url`. Reimplement using native functionality and other methods in `lib/url`, dropping both `url` and `lodash`.
* Split out corresponding tests into their own file too. Drop `chai` and use `jest` expectations exclusively.
* Add a bunch more tests.

#### Testing instructions

The unit tests for this method are pretty extensive, so they should be enough in ensuring that there are no undesirable behaviour changes.

### CODE REVIEW

File: client/lib/url/test/omit-url-params.js
Comment (sirreal) [inline client/lib/url/test/omit-url-params.js]:
```suggestion
			expect( actual ).toBeNull();
```

File: client/lib/url/test/omit-url-params.js
Comment (sirreal) [inline client/lib/url/test/omit-url-params.js]: For primitives, I think we should prefer strict equality `toBe`:

```suggestion
				expect( actual ).toBe( expected );
```

File: client/lib/url/omit-url-params.ts
Comment (sirreal) [inline client/lib/url/omit-url-params.ts]: Strange that we need this type assertion, I'd expect [`typeof` to work as a type guard](https://www.typescriptlang.org/docs/handbook/advanced-types.html#typeof-type-guards).

File: client/lib/url/omit-url-params.ts
Comment (sirreal) [inline client/lib/url/omit-url-params.ts]: What happens with malformed or relative URLs? I'd expect this to throw. Is that the case and is it a change in behavior? Will you add some tests around that?

File: client/lib/url/test/omit-url-params.js
Comment (sgomes) [reply client/lib/url/test/omit-url-params.js]: 👍

File: client/lib/url/test/omit-url-params.js
Comment (sgomes) [reply client/lib/url/test/omit-url-params.js]: 👍

File: client/lib/url/omit-url-params.ts
Comment (sgomes) [reply client/lib/url/omit-url-params.ts]: Looks like it isn't necessary at all. Must have been left behind from when I was trying out different options to handle this. Removed 👍

File: client/lib/url/omit-url-params.ts
Comment (sgomes) [reply client/lib/url/omit-url-params.ts]: Thank you for pointing that out! It seems that node's `parse` is extremely lenient here, and will happily handle things that aren't URLs, as long as there's a `?foo=bar` in there. Ah, how fun.

In order to try to get compatibility as high as possible, I'll need to place this PR on hold while I work on some new additions to `lib/url`.

**Comment by sgomes** [comment]
Rebased on `master`. Note that the build will fail for now, until #37068 gets reviewed and merged, adding the missing method that the implementation relies on.

File: client/lib/url/omit-url-params.ts
Comment (sgomes) [reply client/lib/url/omit-url-params.ts]: The current code should handle any valid URL, whether absolute or having any level of relative-ness. The tests have been expanded to ensure that.

Note that the PR will fail for now, as it's dependent on #37068.

**Comment by sgomes** [comment]
Now that the new methods are available in `lib/url`, this PR is ready to be reviewed once more! 👍

File: client/lib/url/omit-url-params.ts
Comment (sirreal) [inline client/lib/url/omit-url-params.ts]: I'll mention that I was considering the following change, opening up the second type parameter because we really don't care what it is in the falsy url case:

```ts
export default function omitUrlParams( url: Falsy, paramsToOmit: any ): null;
```

While that nicely expresses "falsy urls return null", it seems to complicate things in other ways 🙂 

For signatures like these, as much consistency as possible in the signatures (like you now have) seems best 👍

File: client/lib/url/omit-url-params.ts
Comment (sirreal) [inline client/lib/url/omit-url-params.ts]: As I mentioned in another comment, this isn't covered by the non-Falsy overload and it should be.

File: client/lib/url/test/omit-url-params.js
Comment (sirreal) [inline client/lib/url/test/omit-url-params.js]: Do you think there should be tests with URL encoded characters?

---

No need to change this, but repetitive tests like these can be written concisely and expressively with [```test.each`table`(…)```](https://jestjs.io/docs/en/api#2-testeach-table-name-fn-timeout).

Here's an example (PR, not landed yet) in Calypso: 

https://github.com/Automattic/wp-calypso/blob/b6dcca5510f72d701d4fb4012a7eec9b2ccb4807/packages/localize-monetary-amount/test/localize-monetary-amount.js#L7-L30

File: client/lib/url/omit-url-params.ts
Comment (sirreal) [inline client/lib/url/omit-url-params.ts]: I tend to prefer const + ternary for this sort of thing, but up to you:

```ts
const toOmit: string[] = typeof paramsToOmit === 'string'
	? [ paramsToOmit ]
	: paramsToOmit;
```

File: client/lib/url/omit-url-params.ts
Comment (sirreal) [inline client/lib/url/omit-url-params.ts]: 🤔 `! toOmit` doesn't look quite right, `toOmit` is _always_ and array, and `! toOmit` should _always_ evaluate to `false`. We'll always hit the RHS of this `||`:

```suggestion
		( [ key ] ) => ! toOmit.includes( key )
```

Was this intended to be an optimization? I believe we could return the string unmodified after the valid URL check and avoid further processing:

```ts
	if ( urlType === URL_TYPE.INVALID ) {
		return null;
	}

	// An empty string or array of paramsToOmit is a no-op (with a valid URL)
	if ( ! paramsToOmit.length ) {
		return url;
	}
```

(`""` can't remove any params, can it?)

File: client/lib/url/omit-url-params.ts
Comment (sirreal) [inline client/lib/url/omit-url-params.ts]: Should this move inside the following conditional block where it's used?

File: client/lib/url/omit-url-params.ts
Comment (sirreal) [inline client/lib/url/omit-url-params.ts]: Tested and falsy is a good check here 👍 (it's a string with no `?` if there are no params)

File: client/lib/url/omit-url-params.ts
Comment (sirreal) [inline client/lib/url/omit-url-params.ts]: 👍 The replacement correctly handles URL encoded characters (apparently thanks to `URLSearchParams`)

**Review by sirreal — DISMISSED** [review]
Looks pretty solid for the most part 👍

File: client/lib/url/omit-url-params.ts
Comment (sgomes) [reply client/lib/url/omit-url-params.ts]: That all sounds very reasonable. I should point out that these types and overloads were already mostly in place. I believe I made some small tweaks to them, but the groundwork was already there, and since it seemed solid, I kept it in place.

File: client/lib/url/omit-url-params.ts
Comment (sgomes) [reply client/lib/url/omit-url-params.ts]: 👍

File: client/lib/url/omit-url-params.ts
Comment (sgomes) [reply client/lib/url/omit-url-params.ts]: Committed your suggestion. Thanks!

File: client/lib/url/omit-url-params.ts
Comment (sgomes) [reply client/lib/url/omit-url-params.ts]: Good idea! Done.

File: client/lib/url/omit-url-params.ts
Comment (sgomes) [reply client/lib/url/omit-url-params.ts]: The `! toOmit` looks like old code that I forgot to remove after I rewrote the logic here. Thanks!

I also added your suggested early check 👍

File: client/lib/url/omit-url-params.ts
Comment (sgomes) [reply client/lib/url/omit-url-params.ts]: Moved it further up, actually, to reuse on previous conditional.

File: client/lib/url/test/omit-url-params.js
Comment (sgomes) [reply client/lib/url/test/omit-url-params.js]: I replaced `banana` with `%E3%83%90%E3%83%8A%E3%83%8A`, the URL-encoded rendering of `バナナ` (banana in Japanese/katakana).

As for the test table, is it possible to use it without losing the meaningful test output paths? E.g. `omitUrlParams() › when a path-relative URL is supplied › when a single omitting param is supplied as a string › should return the URL with that param removed`

**Comment by sgomes** [comment]
Sorry for the long delay, @sirreal! All comments addressed.

**Comment by sgomes** [comment]
E2E and desktop failures are unrelated.

**Comment by sgomes** [comment]
Any chance of a final review, @sirreal? This is the last step in getting `lib/url` fully code-split.

**Review by sirreal — APPROVED** [review]
This seems to be in a good place 👍

**Comment by sgomes** [comment]
Thank you for the review, @sirreal! I'll hold off on merging this until early next week, as I'm somewhat concerned it may introduce some IE11 errors (see #38210), as it's introducing more `determineUrlType` usage.

**Pull request merged**
